drivers/platform/maxim/maxim_irq.c

```diff
@@ -44,15 +44,158 @@
+	void *handle;
+	void (*callback)(void *context);
+	void *ctx;
+};
+
+struct event_list {
+	enum no_os_irq_event event;
+	struct no_os_list_desc *actions;
+};
+
+static struct event_list _events[] = {
+	{.event = NO_OS_EVT_GPIO},
+	{.event = NO_OS_EVT_UART_TX_COMPLETE},
+	{.event = NO_OS_EVT_UART_RX_COMPLETE},
+	{.event = NO_OS_EVT_UART_ERROR},
+	{.event = NO_OS_EVT_RTC},
+};
 
 /******************************************************************************/
 /************************ Functions Definitions *******************************/
 /******************************************************************************/
 
+int32_t irq_action_cmp(void *data1, void *data2)
+{
+	return (int32_t)((struct irq_action *)data1)->handle -
+	       (int32_t)((struct irq_action *)data2)->handle;
+}
+
+int32_t irq_action_gpio_cmp(void *data1, void *data2)
+{
+	no_os_gpio_desc *cfg1 = ((struct irq_action *)data1)->handle;
+	no_os_gpio_desc *cfg2 = ((struct irq_action *)data2)->handle;
+
+	if (cfg1->port == cfg2->port && cfg1->number == cfg2->number)
+		return 0;
+	return -1;
+}
+
+static void gpio_irq_callback(void *cbdata)
+{
+	int ret;
+	struct event_list *ee = &_events[NO_OS_EVT_GPIO];
+	struct irq_action *action;
+	struct irq_action key = {.handle = cbdata};
+
+	ret = no_os_list_read_find(ee->actions, (void **)&action, &key);
+	if (ret)
+		return ret;
+
+	if (action->callback)
+		action->callback(action->ctx);
+}
+
+void UART0_IRQHandler()
+{
+	MXC_UART_AsyncHandler(MXC_UART0);
+}
+
+#ifdef MXC_UART1
+void UART1_IRQHandler()
+{
+	MXC_UART_AsyncHandler(MXC_UART1);
+}
+#endif
+
+#ifdef MXC_UART2
+void UART2_IRQHandler()
+{
+	MXC_UART_AsyncHandler(MXC_UART2);
+}
+#endif
+
+#ifdef MXC_UART3
+void UART3_IRQHandler()
+{
+	MXC_UART_AsyncHandler(MXC_UART3);
+}
+#endif
+
+void GPIO0_IRQHandler()
+{
+	MXC_GPIO_Handler(0);
+}
+
+#ifdef MXC_GPIO1
+void GPIO1_IRQHandler()
+{
+	MXC_GPIO_Handler(1);
+}
+#endif
+
+#ifdef MXC_GPIO2
+void GPIO2_IRQHandler()
+{
+	MXC_GPIO_Handler(2);
+}
+#endif
+
+void RTC_IRQHandler()
+{
+	int ret;
+	uint32_t flags = MXC_RTC_GetFlags();
+	struct event_list *evt_list = &_events[NO_OS_EVT_RTC];
+	struct irq_action *action;
+	struct irq_action key = {.handle = MXC_RTC};
+
+	if (flags & MXC_RTC_INT_FL_LONG) {
+		MXC_RTC_ClearFlags(MXC_RTC_INT_FL_LONG);
+		ret = no_os_list_read_find(evt_list->actions, (void **)&action, &key);
+		if (ret < 0)
+			return;
+
+		if (action->callback)
+			action->callback(action->ctx);
+	}
+}
+
+void max_uart_callback(mxc_uart_req_t *req, int result)
+{
+	struct event_list *ee;
+	struct irq_action *a;
+	struct irq_action key = {.handle = req->uart};
+	int ret;
+
+	if (result)
+		ee = &_events[NO_OS_EVT_UART_ERROR];
+	else if (req->txLen == req->txCnt && req->txLen != 0)
+		ee = &_events[NO_OS_EVT_UART_TX_COMPLETE];
+	else
+		ee = &_events[NO_OS_EVT_UART_RX_COMPLETE];
+
+	ret = no_os_list_read_find(ee->actions, (void **)&a, &key);
+	if (ret)
+		return ret;
+	if (a->callback)
+		a->callback(a->ctx);
```

Comment:
The callbacks are added in a list, as far as I can tell. Is this so that there can be more, even on the same event? If so, should we make sure to cal them all?

Reply:
There can be multiple callbacks for the same event, but there should be only one for each peripheral (e.g TX_COMPLETE for UART0 and UART1). In this case, it doesn't make sense to call all of them (the peripheral is selected based on a handler in the same function). But, as you pointed out there might be a problem if the user registers multiple callbacks on the same event/peripheral instance, since only the first would get called.